neqo-transport/src/connection.rs

```diff
@@ -3694,4 +3734,181 @@ mod tests {
     }
+
+    fn check_is_packet_drop(peer: &mut Connection, pkt: Option<Datagram>, dropped: u64, dups: u64) {
+        let dropped_before = peer.stats.dropped_rx;
+        let dups_before = peer.stats.dups_rx;
+        let out = peer.process(pkt, now());
+        assert!(out.as_dgram_ref().is_none());
+        assert_eq!(dropped, peer.stats.dropped_rx - dropped_before);
+        assert_eq!(dups, peer.stats.dups_rx - dups_before);
+    }
+
+    #[test]
+    fn test_discarded_initial_keys() {
+        qdebug!("---- client: generate CH");
+        let mut client = default_client();
+        let init_pkt_c = client.process(None, now());
+        assert!(init_pkt_c.as_dgram_ref().is_some());
+        assert_eq!(init_pkt_c.as_dgram_ref().unwrap().len(), 1232);
+
+        qdebug!("---- server: CH -> SH, EE, CERT, CV, FIN");
+        let mut server = default_server();
+        let init_pkt_s = server.process(Some(init_pkt_c.as_dgram_ref().unwrap().clone()), now());
+        assert!(init_pkt_s.as_dgram_ref().is_some());
+
+        qdebug!("---- client: cert verification");
+        let out = client.process(Some(init_pkt_s.as_dgram_ref().unwrap().clone()), now());
+        assert!(out.as_dgram_ref().is_some());
+
+        // The client has received handshake packet. It will remove the Initial keys.
+        // We will check this by processing init_pkt_s a second time.
+        // The initial packet should be dropped. The packet contains a Handshake packet as well, which
+        // will be marked as dup.
+        check_is_packet_drop(
+            &mut client,
+            Some(init_pkt_s.as_dgram_ref().unwrap().clone()),
+            1,
+            1,
+        );
+
+        assert!(maybe_authenticate(&mut client));
+
+        // The server has not removed the Initial keys yet, because it has not yet received a Handshake
+        // packet from the client.
+        // We will check this by processing init_pkt_c a second time.
+        // The dropped packet is padding. The Initial packet has been mark dup.
+        check_is_packet_drop(
+            &mut server,
+            Some(init_pkt_c.as_dgram_ref().unwrap().clone()),
+            1,
+            1,
+        );
+
+        qdebug!("---- client: SH..FIN -> FIN");
+        let out = client.process(None, now());
+        assert!(out.as_dgram_ref().is_some());
+
+        // The server will process the first Handshake packet.
+        // After this the Initial keys will be dropped.
+        let out = server.process(out.dgram(), now());
+        assert!(out.as_dgram_ref().is_some());
+
+        // Check that the Initial keys are dropped at the server
+        // We will check this by processing init_pkt_c a third time.
+        // The Initial packet has been dropped and padding that follows it.
+        // There is no dups, everything has been dropped.
+        check_is_packet_drop(&mut server, init_pkt_c.dgram(), 1, 0);
+    }
+
+    #[test]
+    fn test_discarded_handshake_keys() {
```

Comment:
As I said before, we should wait for HANDSHAKE_DONE before writing tests for this.  Otherwise, we're in undefined territory.